### PR TITLE
Investments Webpage Box Content and Modal

### PR DESCRIPTION
**Pull request by cindyzhang12**
[//]: # "These comments meant for your reference, they are invisible and don't need to be deleted"

## What's new in this PR?

[//]: # '#####################################'
[//]: # '####### YOUR ANSWER BELOW ###########'
[//]: # '############## END ##################'

### How to Review

[//]: # 'The order in which to review files and 
what to expect when testing locally'
[//]: # '#####################################'
[//]: # '####### YOUR ANSWER BELOW ###########'
[//]: # '############## END ##################'
- Read through Investments.js and Investments.css 
- Implemented the modal for changing dividend preferences, which can be found in the files mentioned above
### Related PRs

[//]: # "Optional - related PRs you're waiting on
/ PRs that will conflict, etc"
[//]: # '#####################################'
[//]: # '####### YOUR ANSWER BELOW ###########'
[//]: # '############## END ##################'
- Need to get this approved so I can start working on the buying shares pages!
### Next steps
- Currently the buy shares button does not work yet and will just redirect to the same page since it hasn't been implemented yet!
[//]: # "What doesn't work yet, what's NOT in this 
PR/has to be done "
[//]: # '#####################################'
[//]: # '####### YOUR ANSWER BELOW ###########'
[//]: # '############## END ##################'

### Tests Performed, Edge Cases

[//]: # '#####################################'
[//]: # '####### YOUR ANSWER BELOW ###########'
[//]: # '############## END ##################'

### Screenshots

[//]: # '#####################################'
[//]: # '#### YOUR SCREENSHOTS BELOW ########'
[//]: # '############## END ##################'

CC: @dfangshuo

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "google-maps-react": "^2.0.2",
     "history": "^4.10.1",
     "react": "^16.10.1",
+    "react-circular-progressbar": "^2.0.3",
```
Comment (dfangshuo) [inline package.json]: actually did not know this existed lmao... nice

File: src/styles/Investments.css [hunk 2]
```diff
@@ -64,3 +70,99 @@
   height: 250px;
   display: flex;
 }
+
+.buy-shares-button {
+  font-family: 'Nunito Sans', sans-serif;
+  justify-content: center;
+}
```
Comment (dfangshuo) [inline src/styles/Investments.css]: this is too general a name. Maybe call it something like investments-buy-shares-button to prevent css collision

File: src/styles/Investments.css [hunk 3]
```diff
@@ -64,3 +70,99 @@
...
+
+.dividend {
+  justify-content: center;
+  align-content: center;
+}
```
Comment (dfangshuo) [inline src/styles/Investments.css]: ditto to above: change it to investsments-dividend-box or something

File: src/styles/Investments.css [hunk 4]
```diff
@@ -64,3 +70,99 @@
...
+
+.buttons {
+  display: flex;
+  flex-direction: column;
+  text-align: center;
+  margin-right: -10px;
+}
```
Comment (dfangshuo) [inline src/styles/Investments.css]: ditto. this one is the most general (need to change)

File: src/styles/Investments.css [hunk 5]
```diff
@@ -64,3 +70,99 @@
...
+
+.investment-box-2 h4 {
+  text-align: left;
+  color: var(--pp-blue);
+}
+
+.preferences {
+  display: flex;
+  flex-direction: column;
+}
```
Comment (dfangshuo) [inline src/styles/Investments.css]: too general

File: src/styles/Investments.css [hunk 6]
```diff
@@ -64,3 +70,99 @@
...
+  text-decoration: none;
+  display: inline-block;
+  font-size: 16px;
+  width: 150px;
+}
+
+.investment-box-2 h4 {
+  text-align: left;
+  color: var(--pp-blue);
+}
```
Comment (dfangshuo) [inline src/styles/Investments.css]: what does 2 mean in this case? maybe more descriptive naming

File: src/styles/Investments.css [hunk 7]
```diff
@@ -64,3 +70,99 @@
...
+.buy-shares-button a {
+  background-color: var(--pp-pink);
+  border-radius: 2.5px;
+  color: white;
+  text-align: center;
+  text-decoration: none;
+  display: inline-block;
+  font-size: 16px;
+  width: 150px;
+}
```
Comment (dfangshuo) [inline src/styles/Investments.css]: investments-buy-shares-button or something

File: src/styles/Investments.css [hunk 8]
```diff
@@ -64,3 +70,99 @@
...
+  display: flex;
+  flex-direction: column;
+}
+
+.preferences h4 {
+  color: var(--pp-black);
+  margin-block-start: 0;
+  margin-block-end: 0;
+  margin-bottom: 8px;
+}
```
Comment (dfangshuo) [inline src/styles/Investments.css]: ditto for everything below. Make naming more specific

File: src/screens/general/Investment.js [hunk 9]
```diff
@@ -1,17 +1,83 @@
...
+                  <CircularProgressbar
+                    viewBox="0 0 0 0"
+                    value={percentage}
+                    text={owner.numberOfShares}
+                    styles={buildStyles({
+                      pathColor: '#cd6795',
+                      textSize: '30px',
+                      textColor: '#cd6795',
+                      trailColor: '#F4F1F24'
+                    })}
```
Comment (dfangshuo) [inline src/screens/general/Investment.js]:
```suggestion
                    styles={buildStyles(circularProgressBarStyles)}
```

Then, define (at the top of this file):

```
const circularProgressBarStyles = {
                      pathColor: '#cd6795',
                      textSize: '30px',
                      textColor: '#cd6795',
                      trailColor: '#F4F1F24'
                    }
```

File: src/screens/general/Investment.js [hunk 10]
```diff
@@ -1,17 +1,83 @@
...
 class Investment extends React.PureComponent {
   render() {
+    const { owner } = this.props;
+    const percentage = owner.numberOfShares * 10;
     return (
       <div className="dashboard">
         <div className="mainheader">
           <h1>My Investment</h1>
           <div className="columnformat">
             <div className="left-content">
```
Comment (dfangshuo) [inline src/screens/general/Investment.js]: what is this class (left-content)? I don't see it defined anywhere

File: src/styles/Investments.css (hunk 6 above)
Comment (dfangshuo) [reply src/styles/Investments.css]: e.g. something like investment-box-top or -bottom would be more descriptive

File: src/screens/general/Investment.js [hunk 11]
```diff
@@ -1,17 +1,196 @@
...
+
+  onSavePreferencesPressed = async () => {
+    const { newIsReceivingDividends } = this.state;
+    this.submitPreference({
+      isReceivingDividends: newIsReceivingDividends
+    });
+  };
+
+  onClickSavePreferences = () => {
+    this.onSavePreferencesPressed();
```
Comment (dfangshuo) [inline src/screens/general/Investment.js]:
```suggestion
    await this.onSavePreferencesPressed();
```

File: src/screens/general/Investment.js [hunk 12]
```diff
@@ -1,17 +1,196 @@
...
+
+  submitPreference = async newIsReceivingDividends => {
+    const { owner } = this.props;
+    await updateOwner(owner.id, newIsReceivingDividends);
+    await refreshUserData(owner.id, owner.isReceivingDividends);
+  };
+
+  onSavePreferencesPressed = async () => {
+    const { newIsReceivingDividends } = this.state;
+    this.submitPreference({
```
Comment (dfangshuo) [inline src/screens/general/Investment.js]:
```suggestion
    await this.submitPreference({
```

File: src/screens/general/Investment.js [hunk 13]
```diff
@@ -1,17 +1,196 @@
...
+
   render() {
+    const { owner } = this.props;
+    const { showModal, newIsReceivingDividends } = this.state;
+    const percentage = owner.numberOfShares * 10;
+    const circularProgressBarStyles = {
+      pathColor: '#cd6795',
+      textSize: '30px',
+      textColor: '#cd6795',
+      trailColor: '#F4F1F24'
```
Comment (dfangshuo) [inline src/screens/general/Investment.js]: define these colors at the top of the file, and call them here. E.g. define

```
const SOME_COLOR = '$cd6795'

...

pathColor: SOME_COLOR

File: src/styles/Investments.css [hunk 14]
```diff
@@ -64,3 +70,150 @@
...
+  transform: scale(0.9);
+}
+
+.dividends-modal {
+  position: absolute;
+  top: 50%;
+  left: 50%;
+  transform: translate(-50%, -50%);
+  border: 0px;
+  background: rgb(255, 255, 255);
```
Comment (dfangshuo) [inline src/styles/Investments.css]:
```suggestion
  background: white;
```

File: src/styles/Investments.css [hunk 15]
```diff
@@ -64,3 +70,150 @@
...
+  background-color: var(--pp-black);
+  border-radius: 40px;
+  color: white;
+  text-align: center;
+  text-decoration: none;
+  font-size: 16px;
+  width: 100px;
+  height: 25px;
+  align-content: flex-end;
+  font-family: 'Nunito Sans', sans-serif;
```
Comment (dfangshuo) [inline src/styles/Investments.css]: same comment as above for fonts

File: src/styles/Investments.css [hunk 16]
```diff
@@ -64,3 +70,150 @@
   height: 250px;
   display: flex;
 }
+
+.investments-buy-shares-button {
+  font-family: 'Nunito Sans', sans-serif;
```
Comment (dfangshuo) [inline src/styles/Investments.css]: is there a reason you are using a custom font here? You shouldn't need to define any type of fonts in these css files

File: src/screens/general/Investment.js [hunk 17]
```diff
@@ -59,11 +59,15 @@ class Investment extends React.PureComponent {
     const { owner } = this.props;
     const { showModal, newIsReceivingDividends } = this.state;
     const percentage = owner.numberOfShares * 10;
+    const progressbarPink = '#cd6795';
+    const progressbarGray = '#F4F1F24';
+    const progressbarText = '30px';
```
Comment (dfangshuo) [inline src/screens/general/Investment.js]: These should be capitalized to denote the fact that their constants:

e.g. PROGRESS_BAR_PINK

File: src/screens/general/Investment.js [hunk 18]
```diff
@@ -1,17 +1,200 @@
...
 
 class Investment extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {
+      showModal: false,
+      newIsReceivingDividends: props.owner.isReceivingDividends
+    };
+
+    this.handleOpenModal = this.handleOpenModal.bind(this);
```
Comment (aivantg) [inline src/screens/general/Investment.js]: We've switched away from using `bind` in our codebase. Use arrow functions instead! Fix is mentioned below

File: src/screens/general/Investment.js [hunk 19]
```diff
@@ -1,17 +1,200 @@
...
+      isReceivingDividends: newIsReceivingDividends
+    });
+  };
+
+  onClickSavePreferences = async () => {
+    await this.onSavePreferencesPressed();
+    this.handleCloseModal();
+  };
+
+  handleOpenModal() {
```
Comment (aivantg) [inline src/screens/general/Investment.js]: to get away with not having to bind this function and the 3 below it, switch this from 

`handleOpenModal() {` 

to 

`handleOpenModal = () => {`

File: src/screens/general/Investment.js [hunk 20]
```diff
@@ -1,17 +1,200 @@
...
+import 'react-circular-progressbar/dist/styles.css';
+import { CircularProgressbar, buildStyles } from 'react-circular-progressbar';
+import { connect } from 'react-redux';
+import { updateOwner } from '../../lib/airtable/request';
+import { refreshUserData } from '../../lib/userDataUtils';
 import '../../styles/Investments.css';
+import GreenCheck from '../../assets/green_check.png';
+import RedX from '../../assets/red_x.png';
 
 class Investment extends React.PureComponent {
```
Comment (aivantg) [inline src/screens/general/Investment.js]: Since this component now has state, you don't want to extend `PureComponent`, just `Component`

File: src/screens/general/Investment.js [hunk 21]
```diff
@@ -1,17 +1,200 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      showModal: false,
+      newIsReceivingDividends: props.owner.isReceivingDividends
```
Comment (aivantg) [inline src/screens/general/Investment.js]: this is good! But the issue here, is if props CHANGES, state won't update because the constructor is only called once. Instead we want to put this in 2 lifecycle methods: `componentDidMount` and `componentDidUpdate`. Frankly this is an edge case we can handle later, but I want to make the comment for the record

File: src/screens/general/Investment.js [hunk 22]
```diff
@@ -1,17 +1,200 @@
...
+    const progressbarPink = '#cd6795';
+    const progressbarGray = '#F4F1F24';
+    const progressbarText = '30px';
+
+    const circularProgressBarStyles = {
+      pathColor: progressbarPink,
+      textSize: progressbarText,
+      textColor: progressbarPink,
+      trailColor: progressbarGray
+    };
```
Comment (aivantg) [inline src/screens/general/Investment.js]: Could we move all these styles to the CircularProgressBar component?

Comment (aivantg) [reply src/screens/general/Investment.js]: oh, i just realized that that wasn't a component we wrote. Something that would be good practice would be isolating the styles for the `CircularProgressBar` in our own component. Something that would look like this: 

You could pass the percentage value down as a prop!
```
import React from 'react';
import { CircularProgressbar, buildStyles } from 'react-circular-progressbar';

class SharesProgressBar extends React.Component { 
  render() {
    const styles = buildStyles(...)
    return (<CircularProgressBar .../>)
  }
}
```

File: src/screens/general/Investment.js [hunk 23]
```diff
@@ -26,4 +209,8 @@ class Investment extends React.PureComponent {
   }
 }
 
-export default Investment;
+const mapStateToProps = state => ({
+  owner: state.userData.owner
```
Comment (aivantg) [inline src/screens/general/Investment.js]: 😍

File: src/screens/general/Investment.js [hunk 24]
```diff
@@ -1,17 +1,200 @@
...
+                </div>
+                <div className="dividends-button">
+                  <button
+                    type="button"
+                    className="change-button"
+                    onClick={this.handleOpenModal}
+                  >
+                    Change
+                  </button>
+                  <Modal
```
Comment (aivantg) [inline src/screens/general/Investment.js]: As this page gets more complicated, we might want to move this modal to it's own component, but again, something we can tackle in the future

File: src/screens/general/Investment.js [hunk 25]
```diff
@@ -1,17 +1,200 @@
...
+  };
+
+  onSavePreferencesPressed = async () => {
+    const { newIsReceivingDividends } = this.state;
+    await this.submitPreference({
+      isReceivingDividends: newIsReceivingDividends
+    });
+  };
+
+  onClickSavePreferences = async () => {
```
Comment (aivantg) [inline src/screens/general/Investment.js]: this function seems kinda extraneous, why not put all of this in `onSavePreferencesPressed`?

File: src/screens/general/Investment.js [hunk 26]
```diff
@@ -1,17 +1,200 @@
...
+    this.state = {
+      showModal: false,
+      newIsReceivingDividends: props.owner.isReceivingDividends
+    };
+
+    this.handleOpenModal = this.handleOpenModal.bind(this);
+    this.handleCloseModal = this.handleCloseModal.bind(this);
+  }
+
+  submitPreference = async newIsReceivingDividends => {
```
Comment (aivantg) [inline src/screens/general/Investment.js]: I think we could consolidate some of these functions!

File: src/screens/general/Investment.js [hunk 27]
```diff
@@ -1,17 +1,200 @@
...
+    const { owner } = this.props;
+    await updateOwner(owner.id, newIsReceivingDividends);
+    await refreshUserData(owner.id, owner.isReceivingDividends);
```
Comment (aivantg) [inline src/screens/general/Investment.js]: Great use of redux!! `refreshUserData` doesn't need to take in the `owner.isReceivingDividends` part though, just the `ownerId`!

File: src/screens/general/Investment.js [hunk 28]
```diff
@@ -1,17 +1,200 @@
...
+                    text={owner.numberOfShares}
+                    styles={buildStyles(circularProgressBarStyles)}
+                  />
+                </div>
+                <div className="box-text">
+                  <h5>
+                    You currently own {owner.numberOfShares} out of 10 possible
+                    shares
+                    <br />
+                    <h4>${owner.numberOfShares * 100}.00</h4>
```
Comment (aivantg) [inline src/screens/general/Investment.js]: don't nest h4 within h5! I think you can just move it below to get the same effect?

File: src/screens/general/Investment.js [hunk 29]
```diff
@@ -1,17 +1,200 @@
...
+                    </a>
+                  </div>
+                  <div className="investments-dividend">Divest</div>
+                </div>
+              </div>
+              <div className="investments-box-dividends">
+                <div className="dividends-preferences-box">
+                  <h4>Dividend Preferences</h4>
+                  <div className="status">
+                    <img
```
Comment (aivantg) [inline src/screens/general/Investment.js]: no idea why...but when you navigate away from this page and then back to it, the size of the check doubles 😅

Maybe we could make the size fixed in the CSS class?

File: src/screens/general/Investment.js (hunk 21 above)
Comment (aivantg) [reply src/screens/general/Investment.js]: this issue is a thing, but it's not as important in this component. The main note is if the user refreshes while on this page, the value of `props.owner.isReceivingDividends` is actually undefined since the owner hasn't been loaded yet. We can talk about this later, but luckily it doesn't have too many implications for the code :)

File: src/screens/general/Investment.js [hunk 30]
```diff
@@ -1,17 +1,200 @@
...
+                        <div>
+                          <input
+                            type="radio"
+                            id="yes-dividends"
+                            name="dividends-choice"
+                            value="yes"
+                            checked={
+                              newIsReceivingDividends ? 'checked' : false
+                            }
+                            onClick={() => this.handleCheckYes()}
```
Comment (aivantg) [inline src/screens/general/Investment.js]: just turn handleCheckYes (and handleCheckNo) into arrow functions, and turn this into `onClick={this.handleCheckYes}`!

File: src/screens/general/Investment.js (hunk 22 above)
Comment (cindyzhang12) [reply src/screens/general/Investment.js]: - note: make new file for this
